### ISSUE TITLE
Data leakage in ragged_preds: test data used for imputation

### ISSUE THREAD

**Issue by SermetPekin** [dhopp1/nowcast_lstm#13]
Hi! Thanks for this useful library. I think I've found a data leakage issue in the example code that may affect results.

When `ragged_preds()` is called with the full dataset (including test data), the default `np.nanmean()` imputation uses statistics from the entire series, including future values:

```python
# Lines 137-138: Model trained on training data only
model = LSTM(training, target, n_timesteps=12, n_models=10)

# Line 156: But ragged_preds gets the full dataset
test_preds = model.ragged_preds(pub_lags, lag, data=data).loc[lambda x: x.date > train_end_date]
#                                            ^^^^
#                                    includes test data!
```

When filling synthetic missingness, `np.nanmean(series)` calculates means from the full `data` series, leaking future information into the evaluation.

## Impact

- Performance metrics (RMSE, MAE) are optimistically biased
- Doesn't replicate real nowcasting conditions

## Proposed Fix

Pre-calculate imputation statistics from training data only:

```python
train_means = training.iloc[:, 1:].mean()

def fill_na_fixed(series):
    if hasattr(series, 'name') and series.name in train_means:
        return train_means[series.name]
    return np.nanmean(series)

model = LSTM(training, target, n_timesteps=12, n_models=10,
             fill_na_func=fill_na_fixed, 
             fill_ragged_edges_func=fill_na_fixed)
```

I've confirmed this on the example dataset. The impact was minimal in this case, but the leakage could significantly affect results with sparser data or different train/test distributions. Happy to submit a PR if helpful.

**Comment by dhopp1** [comment]
Thanks for looking at this! I had a deeper look into this possibility, and I do not believe this is the case; data from the test set is not leaking into the ragged_preds function via the nanmean function.

Why I do not believe this is the case:

- the [`model.ragged_preds`](https://github.com/dhopp1/nowcast_lstm/blob/9c8a84b6075a9f4bbf135595cd3782872e4160c3/nowcast_lstm/LSTM.py#L296 ) function calls the [`model.gen_ragged_X`](https://github.com/dhopp1/nowcast_lstm/blob/9c8a84b6075a9f4bbf135595cd3782872e4160c3/nowcast_lstm/LSTM.py#L309) function. This is the step where the potential leakage may be occurring.
- the [`model.gen_ragged_X`](https://github.com/dhopp1/nowcast_lstm/blob/9c8a84b6075a9f4bbf135595cd3782872e4160c3/nowcast_lstm/LSTM.py#L227) function is defined as calling the the [`data_setup.gen_ragged_X`](https://github.com/dhopp1/nowcast_lstm/blob/9c8a84b6075a9f4bbf135595cd3782872e4160c3/nowcast_lstm/data_setup.py#L289) with some parameters set. Specifically regarding this, it is called with the `other_dataset` parameter set to `self.other_dataset`. The [`other_dataset`](https://github.com/dhopp1/nowcast_lstm/blob/9c8a84b6075a9f4bbf135595cd3782872e4160c3/nowcast_lstm/data_setup.py#L314) parameter exists specifically to deal with this case, where the fillna value is to be calculated via a stipulated other dataset to avoid data leakage, i.e., the training dataset. `self.other_dataset`contains the training data of the model only.

I further confirmed this by running this code:

```
import pandas as pd
import numpy as np

from nowcast_lstm.LSTM import LSTM


# dummy data creation
np.random.seed(42)
dates = pd.date_range(start='2010-01-01', periods=100, freq='MS')
random_data = np.random.rand(100, 4)
df = pd.DataFrame(random_data, columns=['one', 'two', 'three', 'four'])
df.insert(0, 'date', dates)

train = df[:80]
test = df

# modelling 
model = LSTM(train, "one", n_timesteps=12, seeds=[0]) # default parameters with 12 timestep history
    
model.train()

# pub lag info
pub_lags = [2,2,2]
lag = 0

X, y = model.gen_ragged_X(pub_lags, lag, train)

# last two entries of the data series are filled with mean
model.gen_ragged_X(pub_lags, lag, train, "2010-01-01", "2016-08-01")[0][:][0]

# identical figure produced for the data series mean when using the test set as well, ergo no data leakage from the test set
model.gen_ragged_X(pub_lags, lag, test, "2010-01-01", "2018-04-01")[0][:][0]

```

Hopefully, this addresses your question. If not, and you find more specifically how data leakage is happening, let me know. Otherwise, I'll close the issue with these comments.

**Comment by SermetPekin** [comment]
Thank you for the detailed explanation! My bad. I misunderstood previously.
